modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/DartClientCodegen.java

```diff
@@ -474,34 +465,62 @@ public String toDefaultValue(Schema schema) {
 
     @Override
     public String getTypeDeclaration(Schema p) {
-        if (ModelUtils.isArraySchema(p)) {
-            ArraySchema ap = (ArraySchema) p;
-            Schema inner = ap.getItems();
-            return instantiationTypes().get("array") + "<" + getTypeDeclaration(inner) + ">";
-        } else if (ModelUtils.isMapSchema(p)) {
-            Schema inner = getAdditionalProperties(p);
-            return instantiationTypes().get("map") + "<String, " + getTypeDeclaration(inner) + ">";
+        Schema<?> schema = ModelUtils.unaliasSchema(this.openAPI, p, importMapping);
+        Schema<?> target = ModelUtils.isGenerateAliasAsModel() ? p : schema;
+        if (ModelUtils.isArraySchema(target)) {
+            Schema<?> items = getSchemaItems((ArraySchema) schema);
+            return getSchemaType(target) + "<" + getTypeDeclaration(items) + ">";
+        } else if (ModelUtils.isMapSchema(target)) {
+            // Note: ModelUtils.isMapSchema(p) returns true when p is a composed schema that also defines
+            // additionalproperties: true
+            Schema<?> inner = getAdditionalProperties(target);
+            if (inner == null) {
+                LOGGER.error("`{}` (map property) does not have a proper inner type defined. Default to type:string", p.getName());
+                inner = new StringSchema().description("TODO default missing map inner type to string");
+                p.setAdditionalProperties(inner);
+            }
+            return getSchemaType(target) + "<String, " + getTypeDeclaration(inner) + ">";
         }
         return super.getTypeDeclaration(p);
     }
 
     @Override
     public String getSchemaType(Schema p) {
-        String type = super.getSchemaType(p);
-        if (typeMapping.containsKey(type)) {
-            return typeMapping.get(type);
+        String openAPIType = super.getSchemaType(p);
+        if (openAPIType == null) {
+            LOGGER.error("No Type defined for Schema " + p);
         }
-        if (languageSpecificPrimitives.contains(type)) {
-            return type;
+        if (typeMapping.containsKey(openAPIType)) {
+            return typeMapping.get(openAPIType);
         }
-        return toModelName(type);
+        if (languageSpecificPrimitives.contains(openAPIType)) {
+            return openAPIType;
+        }
+        return toModelName(openAPIType);
     }
 
     @Override
     public Map<String, Object> postProcessModels(Map<String, Object> objs) {
         return postProcessModelsEnum(objs);
     }
 
+    @Override
+    public CodegenOperation fromOperation(String path, String httpMethod, Operation operation, List<Server> servers) {
+        final CodegenOperation op = super.fromOperation(path, httpMethod, operation, servers);
```

Comment:
can you add unit tests for this method? is it worth to make it more dart-dio specific as the rest of the changes are dart-dio?

Reply:
Could possibly be moved to the dart-dio generator.

Reply:
makes sense to keep it here for all darts.